rust/lance-arrow/src/lib.rs

```diff
@@ -930,12 +931,124 @@ fn merge_list_struct(left: &dyn Array, right: &dyn Array) -> Arc<dyn Array> {
     }
 }
 
+/// Helper function to merge validity buffers from two struct arrays
+/// Returns None only if both arrays are null at the same position
+///
+/// Special handling for placeholder structs (all-null validity)
+fn merge_struct_validity(
+    left_validity: Option<&arrow_buffer::NullBuffer>,
+    right_validity: Option<&arrow_buffer::NullBuffer>,
+) -> Option<arrow_buffer::NullBuffer> {
+    match (left_validity, right_validity) {
+        // Fast paths: no computation needed
+        (None, None) => None,
+        (Some(left), None) => {
+            // Check if left is a placeholder (all null)
+            if left.null_count() == left.len() {
+                None
+            } else {
+                Some(left.clone())
+            }
+        }
+        (None, Some(right)) => {
+            // Check if right is a placeholder (all null)
+            if right.null_count() == right.len() {
+                None
+            } else {
+                Some(right.clone())
+            }
```

Comment:
I'm not sure if I handled this correctly; it's quite tricky. It seems like lance will generate an all-null validity while the data doesn't exist on this side.

This trick is solely for making test `test_insert_nested_subschemas` pass.

Can you @westonpace double check this test to ensure I didn't understand it in wrong?

Reply:
I think the logic is correct but you could maybe simplify it by dealing with placeholders first.

```
fn normalize_validity(validity: Option<&NullBuffer>) -> Option<&NullBuffer> {
  validity.and_then(|v| {
    if v.null_count() == v.len() {
      None
    } else {
      v
    }
  }
}

fn merge_struct_validity(
    left_validity: Option<&arrow_buffer::NullBuffer>,
    right_validity: Option<&arrow_buffer::NullBuffer>,
) -> Option<arrow_buffer::NullBuffer> {
  left_validity = normalize_validity(left_validity);
  right_validity = normalize_validity(right_validity);
  ...
```